### PR TITLE
Allow implicit creation of production network nodes.

### PR DESCRIPTION
**Pull request by atgeirr**
Nodes with default properties do not need any NODEPROP entries and can be implicitly created by being referred to in BRANPROP. This fixes a bug where such cases would cause an exception.

Creating as draft until test case has been added.

### CODE REVIEW

File: src/opm/input/eclipse/Schedule/Network/ExtNetwork.cpp [hunk 1]
```diff
@@ -75,11 +75,13 @@ const Node& ExtNetwork::root() const {
...
 void ExtNetwork::add_branch(Branch branch)
 {
-    if (!this->has_indexed_node_name(branch.downtree_node())) {
-        this->add_indexed_node_name(branch.downtree_node());
-    }
-    if (!this->has_indexed_node_name(branch.uptree_node())) {
-        this->add_indexed_node_name(branch.uptree_node());
+    for (const std::string& nodename : { branch.downtree_node(), branch.uptree_node() }) {
+        if (!this->has_node(nodename)) {
+            this->m_nodes.insert_or_assign(nodename, Node{nodename});
```
Comment (bska) [inline src/opm/input/eclipse/Schedule/Network/ExtNetwork.cpp:80]: It **probably** doesn't matter, but it's slightly faster to construct the `Node` in place within `m_nodes` than to construct a temporary and then moving it into the `m_nodes` container.  I.e.,
```C++
this->m_nodes.emplace(std::piecewise_construct, std::forward_as_tuple(nodename), std::forward_as_tuple(nodename));
```

Comment (atgeirr) [reply src/opm/input/eclipse/Schedule/Network/ExtNetwork.cpp:80]: Point taken. I must admit I find it quite hard to read though, and hope it is ok to reserve this for more performance sensitive parts. (This insertion will only be executed max twice per BRANPROP record.)

**Comment by atgeirr** [comment]
jenkins build this please

File: src/opm/input/eclipse/Schedule/Network/ExtNetwork.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (atgeirr) [reply src/opm/input/eclipse/Schedule/Network/ExtNetwork.cpp:80]: No, add_node() is intended to be called after all branches have been created. I looked at the possibility but decided against refactoring add_node() to make that work.

[the remark this answers is not in the document]
Comment (GitPaean) [reply src/opm/input/eclipse/Schedule/Network/ExtNetwork.cpp:80]: Then `add_node` is probably a bad name now? It is more likely to be `update` or `assign`?

[the remark this answers is not in the document]
Comment (GitPaean) [reply src/opm/input/eclipse/Schedule/Network/ExtNetwork.cpp:80]: A further question, can we have a fixed pressure node while it does not belong to a branch?

[the remark this answers is not in the document]
Comment (atgeirr) [reply src/opm/input/eclipse/Schedule/Network/ExtNetwork.cpp:80]: I agree add_node is misleading, I will rename it to update_node.

> A further question, can we have a fixed pressure node while it does not belong to a branch?

I think not. All nodes must be connected to a branch.

**Comment by atgeirr** [comment]
jenkins build this please

**Comment by GitPaean** [comment]
Thanks for the update. 

In the new function `update_node` should we make sure the node is there already and we are for sure not creating/inserting a new entry in this function and only update the property of the node?

**Comment by atgeirr** [comment]
> In the new function `update_node` should we make sure the node is there already and we are for sure not creating/inserting a new entry in this function and only update the property of the node?

I think this is guaranteed now, since we check the branches for the node name, and the node will already exist if that check passes. We could simplify it a little, but I wanted to minimize changes for now.

**Comment by GitPaean** [comment]
> I think this is guaranteed now, since we check the branches for the node name, and the node will already exist if that check passes. We could simplify it a little, but I wanted to minimize changes for now.

I understand the point of minimizing changes, but the grantee is ensured through another function. I am okay with the current way though.

**Comment by GitPaean** [comment]
I think the PR is doing its intended work. At the same time, this part of the code will evolve with future testing/cases and new understanding. I am merging this PR now.

**Pull request merged**